File: typescript/tests/metagrammar/characterClass.test.ts
import { MetaGrammar, Parser, CharSet, SyntaxError } from '../../src/index.js';

describe('CharSet Terminal - Direct Construction', () => {
  test('CharSet.range matches characters in range', () => {
    const charSet = CharSet.range('a', 'z');
    const rules = new Map([['S', charSet]]);

    let parser = new Parser({ topRuleName: 'S', rules, input: 'a' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);

    parser = new Parser({ topRuleName: 'S', rules, input: 'm' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);

    parser = new Parser({ topRuleName: 'S', rules, input: 'z' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);

    parser = new Parser({ topRuleName: 'S', rules, input: 'A' });
    expect(parser.parse().root instanceof SyntaxError).toBe(true);
  });

  test('CharSet.char matches single character', () => {
    const charSet = CharSet.char('x');
    const rules = new Map([['S', charSet]]);

    let parser = new Parser({ topRuleName: 'S', rules, input: 'x' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);

    parser = new Parser({ topRuleName: 'S', rules, input: 'y' });
    expect(parser.parse().root instanceof SyntaxError).toBe(true);
  });

  test('CharSet with multiple ranges', () => {
    // [a-zA-Z0-9]
    const charSet = new CharSet([
      ['a'.charCodeAt(0), 'z'.charCodeAt(0)],
      ['A'.charCodeAt(0), 'Z'.charCodeAt(0)],
      ['0'.charCodeAt(0), '9'.charCodeAt(0)],
    ]);
    const rules = new Map([['S', charSet]]);

    // Test lowercase
    let parser = new Parser({ topRuleName: 'S', rules, input: 'a' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);
    parser = new Parser({ topRuleName: 'S', rules, input: 'z' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);

    // Test uppercase
    parser = new Parser({ topRuleName: 'S', rules, input: 'A' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);
    parser = new Parser({ topRuleName: 'S', rules, input: 'Z' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);

    // Test digits
    parser = new Parser({ topRuleName: 'S', rules, input: '0' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);
    parser = new Parser({ topRuleName: 'S', rules, input: '9' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);

    // Test non-alphanumeric
    parser = new Parser({ topRuleName: 'S', rules, input: '!' });
    expect(parser.parse().root instanceof SyntaxError).toBe(true);
    parser = new Parser({ topRuleName: 'S', rules, input: ' ' });
    expect(parser.parse().root instanceof SyntaxError).toBe(true);
  });

  test('CharSet with inversion', () => {
    // [^a-z] - matches anything NOT a lowercase letter
    const charSet = new CharSet(
      [['a'.charCodeAt(0), 'z'.charCodeAt(0)]],
      true
    );
    const rules = new Map([['S', charSet]]);

    // Should NOT match lowercase
    let parser = new Parser({ topRuleName: 'S', rules, input: 'a' });
    expect(parser.parse().root instanceof SyntaxError).toBe(true);
    parser = new Parser({ topRuleName: 'S', rules, input: 'm' });
    expect(parser.parse().root instanceof SyntaxError).toBe(true);
    parser = new Parser({ topRuleName: 'S', rules, input: 'z' });
    expect(parser.parse().root instanceof SyntaxError).toBe(true);

    // Should match uppercase, digits, symbols
    parser = new Parser({ topRuleName: 'S', rules, input: 'A' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);
    parser = new Parser({ topRuleName: 'S', rules, input: '5' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);
    parser = new Parser({ topRuleName: 'S', rules, input: '!' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);
  });

  test('CharSet with inverted multiple ranges', () => {
    // [^a-zA-Z] - matches anything NOT a letter
    const charSet = new CharSet(
      [
        ['a'.charCodeAt(0), 'z'.charCodeAt(0)],
        ['A'.charCodeAt(0), 'Z'.charCodeAt(0)],
      ],
      true
    );
    const rules = new Map([['S', charSet]]);

    // Should NOT match letters
    let parser = new Parser({ topRuleName: 'S', rules, input: 'a' });
    expect(parser.parse().root instanceof SyntaxError).toBe(true);
    parser = new Parser({ topRuleName: 'S', rules, input: 'Z' });
    expect(parser.parse().root instanceof SyntaxError).toBe(true);

    // Should match digits and symbols
    parser = new Parser({ topRuleName: 'S', rules, input: '5' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);
    parser = new Parser({ topRuleName: 'S', rules, input: '!' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);
  });

  test('CharSet.notRange convenience constructor', () => {
    const charSet = CharSet.notRange('0', '9');
    const rules = new Map([['S', charSet]]);

    // Should NOT match digits
    let parser = new Parser({ topRuleName: 'S', rules, input: '5' });
    expect(parser.parse().root instanceof SyntaxError).toBe(true);

    // Should match non-digits
    parser = new Parser({ topRuleName: 'S', rules, input: 'a' });
    expect(parser.parse().root instanceof SyntaxError).toBe(false);
  });

  test('CharSet toString formats correctly', () => {
    expect(CharSet.range('a', 'z').toString()).toBe('[a-z]');
    expect(CharSet.char('x').toString()).toBe('[x]');
    expect(
      new CharSet([
        ['a'.charCodeAt(0), 'z'.charCodeAt(0)],
        ['0'.charCodeAt(0), '9'.charCodeAt(0)],
      ]).toString()
    ).toBe('[a-z0-9]');
    expect(
      new CharSet([['a'.charCodeAt(0), 'z'.charCodeAt(0)]], true).toString()
    ).toBe('[^a-z]');
  });

  test('CharSet handles empty input', () => {
    const charSet = CharSet.range('a', 'z');
    const rules = new Map([['S', charSet]]);

    const parser = new Parser({ topRuleName: 'S', rules, input: '' });
    expect(parser.parse().root instanceof SyntaxError).toBe(true);
  });
});

describe('MetaGrammar - Character Classes', () => {
  test('simple character range', () => {
    const grammar = `
      Digit <- [0-9];
    `;

    const rules = MetaGrammar.parseGrammar(grammar);

    let parser = new Parser({ topRuleName: 'Digit', rules, input: '5' });
    let parseResult = parser.parse();
    let result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(false);
    expect(result.len).toBe(1);

    parser = new Parser({ topRuleName: 'Digit', rules, input: 'a' });
    parseResult = parser.parse();
    result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(true);
  });

  test('multiple character ranges', () => {
    const grammar = `
      AlphaNum <- [a-zA-Z0-9];
    `;

    const rules = MetaGrammar.parseGrammar(grammar);

    let parser = new Parser({ topRuleName: 'AlphaNum', rules, input: 'a' });
    let parseResult = parser.parse();
    let result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(false);

    parser = new Parser({ topRuleName: 'AlphaNum', rules, input: 'Z' });
    parseResult = parser.parse();
    result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(false);

    parser = new Parser({ topRuleName: 'AlphaNum', rules, input: '5' });
    parseResult = parser.parse();
    result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(false);

    parser = new Parser({ topRuleName: 'AlphaNum', rules, input: '!' });
    parseResult = parser.parse();
    result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(true);
  });

  test('character class with individual characters', () => {
    const grammar = `
      Vowel <- [aeiou];
    `;

    const rules = MetaGrammar.parseGrammar(grammar);

    let parser = new Parser({ topRuleName: 'Vowel', rules, input: 'a' });
    let parseResult = parser.parse();
    let result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(false);

    parser = new Parser({ topRuleName: 'Vowel', rules, input: 'e' });
    parseResult = parser.parse();
    result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(false);

    parser = new Parser({ topRuleName: 'Vowel', rules, input: 'b' });
    parseResult = parser.parse();
    result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(true);
  });

  test('negated character class', () => {
    const grammar = `
      NotDigit <- [^0-9];
    `;

    const rules = MetaGrammar.parseGrammar(grammar);

    let parser = new Parser({ topRuleName: 'NotDigit', rules, input: 'a' });
    let parseResult = parser.parse();
    let result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(false);

    parser = new Parser({ topRuleName: 'NotDigit', rules, input: '5' });
    parseResult = parser.parse();
    result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(true);
  });

  test('escaped characters in character class', () => {
    const grammar = String.raw`
      Special <- [\t\n];
    `;

    const rules = MetaGrammar.parseGrammar(grammar);

    let parser = new Parser({ topRuleName: 'Special', rules, input: '\t' });
    let parseResult = parser.parse();
    let result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(false);

    parser = new Parser({ topRuleName: 'Special', rules, input: '\n' });
    parseResult = parser.parse();
    result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(false);

    parser = new Parser({ topRuleName: 'Special', rules, input: ' ' });
    parseResult = parser.parse();
    result = parseResult.root;
    expect(result instanceof SyntaxError).toBe(true);
  });
});

File: dart/test/metagrammar/character_class_test.dart
import 'package:squirrel_parser/squirrel_parser.dart';
import 'package:test/test.dart';

void main() {
  group('CharSet Terminal - Direct Construction', () {
    test('CharSet.range matches characters in range', () {
      final charSet = CharSet.range('a', 'z');
      final rules = {'S': charSet};

      var parser = Parser(topRuleName: 'S', rules: rules, input: 'a');
      expect(parser.parse().root is! SyntaxError, isTrue, reason: 'should match a');

      parser = Parser(topRuleName: 'S', rules: rules, input: 'm');
      expect(parser.parse().root is! SyntaxError, isTrue, reason: 'should match m');

      parser = Parser(topRuleName: 'S', rules: rules, input: 'z');
      expect(parser.parse().root is! SyntaxError, isTrue, reason: 'should match z');

      parser = Parser(topRuleName: 'S', rules: rules, input: 'A');
      expect(parser.parse().root is SyntaxError, isTrue, reason: 'should not match A');
    });

    test('CharSet.char matches single character', () {
      final charSet = CharSet.char('x');
      final rules = {'S': charSet};

      var parser = Parser(topRuleName: 'S', rules: rules, input: 'x');
      expect(parser.parse().root is! SyntaxError, isTrue, reason: 'should match x');

      parser = Parser(topRuleName: 'S', rules: rules, input: 'y');
      expect(parser.parse().root is SyntaxError, isTrue, reason: 'should not match y');
    });

    test('CharSet with multiple ranges', () {
      // [a-zA-Z0-9]
      final charSet = CharSet([
        ('a'.codeUnitAt(0), 'z'.codeUnitAt(0)),
        ('A'.codeUnitAt(0), 'Z'.codeUnitAt(0)),
        ('0'.codeUnitAt(0), '9'.codeUnitAt(0)),
      ]);
      final rules = {'S': charSet};

      // Test lowercase
      var parser = Parser(topRuleName: 'S', rules: rules, input: 'a');
      expect(parser.parse().root is! SyntaxError, isTrue);
      parser = Parser(topRuleName: 'S', rules: rules, input: 'z');
      expect(parser.parse().root is! SyntaxError, isTrue);

      // Test uppercase
      parser = Parser(topRuleName: 'S', rules: rules, input: 'A');
      expect(parser.parse().root is! SyntaxError, isTrue);
      parser = Parser(topRuleName: 'S', rules: rules, input: 'Z');
      expect(parser.parse().root is! SyntaxError, isTrue);

      // Test digits
      parser = Parser(topRuleName: 'S', rules: rules, input: '0');
      expect(parser.parse().root is! SyntaxError, isTrue);
      parser = Parser(topRuleName: 'S', rules: rules, input: '9');
      expect(parser.parse().root is! SyntaxError, isTrue);

      // Test non-alphanumeric
      parser = Parser(topRuleName: 'S', rules: rules, input: '!');
      expect(parser.parse().root is SyntaxError, isTrue);
      parser = Parser(topRuleName: 'S', rules: rules, input: ' ');
      expect(parser.parse().root is SyntaxError, isTrue);
    });

    test('CharSet with inversion', () {
      // [^a-z] - matches anything NOT a lowercase letter
      final charSet = CharSet([
        ('a'.codeUnitAt(0), 'z'.codeUnitAt(0)),
      ], inverted: true);
      final rules = {'S': charSet};

      // Should NOT match lowercase
      var parser = Parser(topRuleName: 'S', rules: rules, input: 'a');
      expect(parser.parse().root is SyntaxError, isTrue, reason: 'should not match a');
      parser = Parser(topRuleName: 'S', rules: rules, input: 'm');
      expect(parser.parse().root is SyntaxError, isTrue, reason: 'should not match m');
      parser = Parser(topRuleName: 'S', rules: rules, input: 'z');
      expect(parser.parse().root is SyntaxError, isTrue, reason: 'should not match z');

      // Should match uppercase, digits, symbols
      parser = Parser(topRuleName: 'S', rules: rules, input: 'A');
      expect(parser.parse().root is! SyntaxError, isTrue, reason: 'should match A');
      parser = Parser(topRuleName: 'S', rules: rules, input: '5');
      expect(parser.parse().root is! SyntaxError, isTrue, reason: 'should match 5');
      parser = Parser(topRuleName: 'S', rules: rules, input: '!');
      expect(parser.parse().root is! SyntaxError, isTrue, reason: 'should match !');
    });

    test('CharSet with inverted multiple ranges', () {
      // [^a-zA-Z] - matches anything NOT a letter
      final charSet = CharSet([
        ('a'.codeUnitAt(0), 'z'.codeUnitAt(0)),
        ('A'.codeUnitAt(0), 'Z'.codeUnitAt(0)),
      ], inverted: true);
      final rules = {'S': charSet};

      // Should NOT match letters
      var parser = Parser(topRuleName: 'S', rules: rules, input: 'a');
      expect(parser.parse().root is SyntaxError, isTrue, reason: 'should not match a');
      parser = Parser(topRuleName: 'S', rules: rules, input: 'Z');
      expect(parser.parse().root is SyntaxError, isTrue, reason: 'should not match Z');

      // Should match digits and symbols
      parser = Parser(topRuleName: 'S', rules: rules, input: '5');
      expect(parser.parse().root is! SyntaxError, isTrue, reason: 'should match 5');
      parser = Parser(topRuleName: 'S', rules: rules, input: '!');
      expect(parser.parse().root is! SyntaxError, isTrue, reason: 'should match !');
    });

    test('CharSet.notRange convenience constructor', () {
      final charSet = CharSet.notRange('0', '9');
      final rules = {'S': charSet};

      // Should NOT match digits
      var parser = Parser(topRuleName: 'S', rules: rules, input: '5');
      expect(parser.parse().root is SyntaxError, isTrue, reason: 'should not match 5');

      // Should match non-digits
      parser = Parser(topRuleName: 'S', rules: rules, input: 'a');
      expect(parser.parse().root is! SyntaxError, isTrue, reason: 'should match a');
    });

    test('CharSet toString formats correctly', () {
      expect(CharSet.range('a', 'z').toString(), equals('[a-z]'));
      expect(CharSet.char('x').toString(), equals('[x]'));
      expect(CharSet([
        ('a'.codeUnitAt(0), 'z'.codeUnitAt(0)),
        ('0'.codeUnitAt(0), '9'.codeUnitAt(0)),
      ]).toString(), equals('[a-z0-9]'));
      expect(CharSet([
        ('a'.codeUnitAt(0), 'z'.codeUnitAt(0)),
      ], inverted: true).toString(), equals('[^a-z]'));
    });

    test('CharSet handles empty input', () {
      final charSet = CharSet.range('a', 'z');
      final rules = {'S': charSet};

      final parser = Parser(topRuleName: 'S', rules: rules, input: '');
      expect(parser.parse().root is SyntaxError, isTrue, reason: 'should not match empty');
    });
  });

  group('MetaGrammar - Character Classes', () {
    test('simple character range', () {
      const grammar = '''
        Digit <- [0-9];
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);

      var parser = Parser(topRuleName: 'Digit', rules: rules, input: '5');
      var parseResult = parser.parse();
      var result = parseResult.root;
      expect(result is! SyntaxError, isTrue, reason: 'should match digit');
      expect(result.len, equals(1));

      parser = Parser(topRuleName: 'Digit', rules: rules, input: 'a');
      parseResult = parser.parse();
      result = parseResult.root;
      expect(result is SyntaxError, isTrue, reason: 'should fail on non-digit');
    });

    test('multiple character ranges', () {
      const grammar = '''
        AlphaNum <- [a-zA-Z0-9];
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);

      var parser = Parser(topRuleName: 'AlphaNum', rules: rules, input: 'a');
      var parseResult = parser.parse();
      var result = parseResult.root;
      expect(result is! SyntaxError, isTrue);

      parser = Parser(topRuleName: 'AlphaNum', rules: rules, input: 'Z');
      parseResult = parser.parse();
      result = parseResult.root;
      expect(result is! SyntaxError, isTrue);

      parser = Parser(topRuleName: 'AlphaNum', rules: rules, input: '5');
      parseResult = parser.parse();
      result = parseResult.root;
      expect(result is! SyntaxError, isTrue);

      parser = Parser(topRuleName: 'AlphaNum', rules: rules, input: '!');
      parseResult = parser.parse();
      result = parseResult.root;
      expect(result is SyntaxError, isTrue,
          reason: 'should fail on non-alphanumeric');
    });

    test('character class with individual characters', () {
      const grammar = '''
        Vowel <- [aeiou];
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);

      var parser = Parser(topRuleName: 'Vowel', rules: rules, input: 'a');
      var parseResult = parser.parse();
      var result = parseResult.root;
      expect(result is! SyntaxError, isTrue);

      parser = Parser(topRuleName: 'Vowel', rules: rules, input: 'e');
      parseResult = parser.parse();
      result = parseResult.root;
      expect(result is! SyntaxError, isTrue);

      parser = Parser(topRuleName: 'Vowel', rules: rules, input: 'b');
      parseResult = parser.parse();
      result = parseResult.root;
      expect(result is SyntaxError, isTrue, reason: 'should fail on consonant');
    });

    test('negated character class', () {
      const grammar = '''
        NotDigit <- [^0-9];
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);

      var parser = Parser(topRuleName: 'NotDigit', rules: rules, input: 'a');
      var parseResult = parser.parse();
      var result = parseResult.root;
      expect(result is! SyntaxError, isTrue);

      parser = Parser(topRuleName: 'NotDigit', rules: rules, input: '5');
      parseResult = parser.parse();
      result = parseResult.root;
      expect(result is SyntaxError, isTrue, reason: 'should fail on digit');
    });

    test('escaped characters in character class', () {
      const grammar = r'''
        Special <- [\t\n];
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);

      var parser = Parser(topRuleName: 'Special', rules: rules, input: '\t');
      var parseResult = parser.parse();
      var result = parseResult.root;
      expect(result is! SyntaxError, isTrue);

      parser = Parser(topRuleName: 'Special', rules: rules, input: '\n');
      parseResult = parser.parse();
      result = parseResult.root;
      expect(result is! SyntaxError, isTrue);

      parser = Parser(topRuleName: 'Special', rules: rules, input: ' ');
      parseResult = parser.parse();
      result = parseResult.root;
      expect(result is SyntaxError, isTrue, reason: 'should fail on space');
    });
  });
}
